### PR TITLE
feat(NODE-5470)!: convert remaining FLE to TS and drop support for `onKMSProvidersRefresh`

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

The remaining FLE files and tests are converted to TS.  Additional changes:

- buffer_pool.js was removed and now FLE code imports the buffer pool from `utils.ts`.
- `once` and `child_process` are lazily imported in the state machine and mongodbcyptd manager, respectively 
- lint has been re-enabled and once again covers all tests and src
- `databaseNamespace` and `collectionNamespace` have been removed in favor of `MongoDBNamespace`.

One other notable change is that when we fail to get kms provider information, the error thrown _used to_ attach an `originalError` property.  I adjusted this to use the error's `cause`, since Node16+ and MongoCryptErrors support the `cause` property.

I also removed support for `onKMSProvidersRefresh` here.

##### Is there new documentation needed for these changes?

No, documentation is coming in a follow up.

### `ClientEncryption.onKMSProvidersRefresh` function removed

`ClientEncryption.onKMSProvidersRefresh` was added as a public API in version 2.3.0 of `mongodb-client-encryption` to allow for automatic refresh of KMS provider credentials.  Subsequently, we've added the capability to automatically refresh KMS credentials using the KMS provider's preferred refresh mechanism, and `onKMSProviderRefresh` is no longer used.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:300]: You'll notice a lot of type casts here.  `schemaMap` explicitly needs to be a buffer for our bindings, but our bson serialize is platform-compatible and is typed to return `Uint8Arrays`.  At runtime, BSON returns `Buffer`s for Node, so it's still valid, but these casts are unavoidable.

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:385]: Typescript complains here, because we (developers) know that `init` is a no-op when `this._bypassMongocryptdAndCryptShared || this.cryptSharedLibVersionInfo` is specified and that if neither of those are true, we always have a _mongocryptdManager` and client.  We can cast it away but I think errors are a good practice, because these provide actionable feedback if the assumption that these properties are always defined when we need to spawn a mongocryptd is ever violated.

Comment (baileympearson) [reply src/client-side-encryption/autoEncrypter.ts:385]: Maybe a `MongoCryptImpossibleState` error?

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:465]: `encrypt` is only called in one location (CryptoConnection.command) and we pass the command arguments in.

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:474]: I decided not to clean up argument checking like this - for our public APIs in the CE object, I think they're worthwhile to keep.  Private APIs like this I think could be cleaned up, but for consistency and fewer changes I left them as-is.

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:503]: note: these TODOs were here before me.

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:509]: perhaps questionable if this is needed .. I think we only need the user's BSON serialization options (if that) but future work can investigate this further.

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:553]: I was unsure how to get around these errors - any thoughts?

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:559]: These will go away after we refactor to async-await, but for now TS complains because technically `callback` could have been re-assigned by something before we execute this code.  In practice, we know it never is re-assigned.

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [inline src/client-side-encryption/autoEncrypter.ts:575]: removed `onKmsProvidersRefresh`.  

We could consider removing this entirely and just calling `loadCredentials` directly in the state machine.

File: src/client-side-encryption/clientEncryption.ts
Comment (baileympearson) [inline src/client-side-encryption/clientEncryption.ts:24]: Types in this file were directly copied in from https://github.com/mongodb/libmongocrypt/blob/8588825112a93145bf82b4b7519070c101c45a43/bindings/node/index.d.ts#L448, with the exception of a few modifications:

- TLS options 
-

File: src/client-side-encryption/clientEncryption.ts
Comment (baileympearson) [inline src/client-side-encryption/clientEncryption.ts:205]: We will remove callback support once this PR is merged - I intentionally did not TSify `promiseOrCallback` or `maybePromise`.

File: src/client-side-encryption/clientEncryption.ts
Comment (baileympearson) [inline src/client-side-encryption/clientEncryption.ts:743]: I refactored options construction here because the old approach gives TS errors.  For example, keyId.

```typescript
const contextOptions: ExplicitEncryptionContextOptions = Object.assign({}, options);
if (options.keyId) {
  contextOptions.keyId = options.keyId.buffer;
}
```

In the above, TS complains in the first line because we assign a `Binary` to the `keyId` option, which accepts a buffer, only to re-assign it in the if statement.  Most of the options were impacted by this issue.

File: src/client-side-encryption/common.js
Comment (baileympearson) [inline src/client-side-encryption/common.js:1]: `debug` was moved to the state machine, and the namespace helpers were removed in favor of `MongoDBNamespace`.  the remainder of the file is untouched and I disabled lint because we plan to remove this entirely in a follow-up-ticket.

File: src/client-side-encryption/stateMachine.ts
Comment (baileympearson) [inline src/client-side-encryption/stateMachine.ts:66]: These three properties are only used in the JS layer and are only added by CE and AE to communicate context to the state machine.  A future refactor could provide them as a "context" object to `StateMachine.execute` but I left as-is for now.

We use declaration merging to add the properties to the `MongoCryptContext` interface only in the driver.

File: src/client-side-encryption/stateMachine.ts
Comment (baileympearson) [inline src/client-side-encryption/stateMachine.ts:125]: This is new.  Previously, we simply passed the `options` object into `serialize` to provide bson options.  I figured it makes sense to pull off all supported BSON serialization options and store them separately for clarity.

File: src/client-side-encryption/stateMachine.ts
Comment (baileympearson) [inline src/client-side-encryption/stateMachine.ts:342]: This is a behavior change.  We now have access to the `cause` property - may as well make proper use of it.

Comment (nbbeeken) [reply src/client-side-encryption/stateMachine.ts:367]: I assume this PR comment is the todo: `// TODO: make note of this`?

File: src/client-side-encryption/stateMachine.ts
Comment (nbbeeken) [reply src/client-side-encryption/stateMachine.ts:151]: SGTM

File: src/client-side-encryption/cryptoCallbacks.ts
Comment (nbbeeken) [inline src/client-side-encryption/cryptoCallbacks.ts:9]: I think we should strive to lower the contract needed here, the APIs accept Uint8Array as readily as they do Buffers. Are there Buffer specific APIs needed here?

File: src/client-side-encryption/mongocryptdManager.ts
Comment (nbbeeken) [inline src/client-side-encryption/mongocryptdManager.ts:60]: It is not exactly questionable, the source of "truth" for a successful launch is a mongocryptd process we can connect to. Since this channel of getting an error can occur at anytime we defer to the connection to mongocryptd being broken instead which is kind of our way of "monitoring" that it is still up.

File: src/client-side-encryption/stateMachine.ts
Comment (nbbeeken) [inline src/client-side-encryption/stateMachine.ts:71]:
```suggestion
function debug(msg: unknown) {
```

Not used elsewhere right?

File: src/client-side-encryption/stateMachine.ts
Comment (nbbeeken) [inline src/client-side-encryption/stateMachine.ts:65]:
```
'tlsDisableOCSPEndpointCheck',
'tlsDisableCertificateRevocationCheck'
```
These are not client options. Maybe we should leave them though but with a comment to help clear up potential confusion.

File: src/client-side-encryption/stateMachine.ts
Comment (nbbeeken) [inline src/client-side-encryption/stateMachine.ts:98]: We should remove these options from documentation

File: src/index.ts
Comment (nbbeeken) [inline src/index.ts:102]: Can you move these down to live with the other type exports? we have a line "type only exports below, these are removed from emitted JS" breaking up the runtime values from the types.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/utils.ts
Comment (baileympearson) [reply src/utils.ts:350]: It has the signature as the parent method it's overriding.

File: src/client-side-encryption/clientEncryption.ts
Comment (baileympearson) [reply src/client-side-encryption/clientEncryption.ts:42]: https://jira.mongodb.org/browse/NODE-5423

File: src/client-side-encryption/mongocryptdManager.ts
Comment (baileympearson) [reply src/client-side-encryption/mongocryptdManager.ts:60]: Right - but we could race `spawn` and `error` together, and once successfully spawned stop listening for errors. 

It's strange that there's no way to know if auto spawn was successful other than a failing auto encrypted command (which manifests itself as a failure to connect the mongocryptd client).

File: src/client-side-encryption/stateMachine.ts
Comment (baileympearson) [reply src/client-side-encryption/stateMachine.ts:462]: I added it as an AC in the ticket

File: src/index.ts
Comment (baileympearson) [reply src/index.ts:102]: Is this something we can enforce (via linting or another measure)?  (if it matters...?)

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [reply src/client-side-encryption/autoEncrypter.ts:388]: it isn't fixable for users if they encounter this error - we create a `mongocryptd` client so if they encounter this error, we have a bug

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [reply src/client-side-encryption/autoEncrypter.ts:477]: Can you clarify?  Why would we remove guards against invalid arguments?

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [reply src/client-side-encryption/autoEncrypter.ts:512]: The types declared allow `CommandOptions` because that's currently what we pass in, but we only should need the BSON serialization options (and even those are questionable).

File: src/client-side-encryption/autoEncrypter.ts
Comment (nbbeeken) [reply src/client-side-encryption/autoEncrypter.ts:502]: I think the argument validation is fine, a little unnecessary if we trust our TS. But I thought "clean up" would include making this MongoError subclass (like InvalidArgument)

File: src/client-side-encryption/autoEncrypter.ts
Comment (nbbeeken) [reply src/client-side-encryption/autoEncrypter.ts:536]: I think this actually the source of that old sessions bug NODE-4896, if we didn't spread here the session wouldn't have made it to the operations. Do we want to make a change here or file a follow up to investigate?

File: src/client-side-encryption/autoEncrypter.ts
Comment (nbbeeken) [inline src/client-side-encryption/autoEncrypter.ts:218]: nit: Can you put the comment into the tsdoc on this variable?

File: src/client-side-encryption/autoEncrypter.ts
Comment (nbbeeken) [inline src/client-side-encryption/autoEncrypter.ts:253]: I don't feel strongly but a wise Bailey once said optional booleans make an enum with 3 values. Any harm in setting this to false?

File: src/client-side-encryption/autoEncrypter.ts
Comment (nbbeeken) [reply src/client-side-encryption/autoEncrypter.ts]: Do we want to make this change?

File: src/client-side-encryption/mongocryptdManager.ts
Comment (nbbeeken) [reply src/client-side-encryption/mongocryptdManager.ts:60]: This might be something that is captured in the specification (or should be). We can discuss the finer points in tech chat but for now let's leave something more descriptive than "perhaps questionable". I'm thinking let's capture the intention regardless of our takes on the approach. 

Something like:
> Add an empty error listener to squash child process errors
> If the process encountered an error we will learn of it when we fail to connect

File: src/client-side-encryption/mongocryptdManager.ts
Comment (nbbeeken) [inline src/client-side-encryption/mongocryptdManager.ts:51]:
```suggestion
    // Spawned with stdio: ignore and detached: true
```

File: src/utils.ts
Comment (nbbeeken) [reply src/utils.ts:350]: Can you clarify? Why wouldn't it be `fromString(namespace?: string)`?

File: src/client-side-encryption/autoEncrypter.ts
Comment (baileympearson) [reply src/client-side-encryption/autoEncrypter.ts:536]: No, the root cause was explicitly passing the session into driver helpers in the StateMachine: https://github.com/mongodb/libmongocrypt/commit/c9e44e7c0842683659e2a068696e6a712c33ab43

I checked in the `StateMachine` and the only four values we're accessing on the options object are the values we pass in here, so I removed the spreading.

File: src/utils.ts
Comment (baileympearson) [reply src/utils.ts:350]: Oh I misunderstood your question.  fixed

File: src/index.ts
Comment (nbbeeken) [reply src/index.ts]: We can look into it, our exports are kept sorted but the "sections" are manually maintained.

File: src/client-side-encryption/autoEncrypter.ts
Comment (nbbeeken) [reply src/client-side-encryption/autoEncrypter.ts]: Seems overkill fair enough

**Review by nbbeeken — APPROVED** [review]

**Pull request merged**
